llvm/lib/CASObjectFormats/FlatV1.cpp

```diff
@@ -436,28 +424,43 @@ Expected<BlockRef> BlockRef::create(CompileUnitBuilder &CUB,
     Fixups.back().Offset = E->getOffset();
   }
 
-  // Normalize content and put it at the end of the block.
-  // FIXME: assume current block alignment is the alignment for padding and just
-  // pad every block.
-  Optional<Align> TrailingNopsAlignment;
-  TrailingNopsAlignment = Align{Block.getAlignment()};
-  StringRef Content(Block.getContent().begin(), Block.getSize());
-  SmallString<1024> MutableContentStorage;
-  if (Error E = helpers::canonicalizeContent(CUB.TT.getArch(), Content,
-                                             MutableContentStorage, Fixups,
-                                             TrailingNopsAlignment)
-                    .moveInto(Content))
-    return E;
-
+  // Create BlockData.
+  SmallString<1024> EncodeContent;
+  if (!Block.isZeroFill()) {
+    SmallString<1024> MutableContentStorage;
+    // Normalize content and put it at the end of the block.
+    // FIXME: assume current block alignment is the alignment for padding and
+    // just pad every block.
+    Optional<Align> TrailingNopsAlignment;
+    TrailingNopsAlignment = Align{Block.getAlignment()};
+    StringRef Content(Block.getContent().begin(), Block.getSize());
+    if (Error E = helpers::canonicalizeContent(CUB.TT.getArch(), Content,
+                                               MutableContentStorage, Fixups,
+                                               TrailingNopsAlignment)
+                      .moveInto(Content))
+      return E;
+    data::BlockData::encode(Content.size(), Block.getAlignment(),
+                            Block.getAlignmentOffset(), Content,
+                            ArrayRef<Fixup>(), EncodeContent);
```

Comment:
That is because my Ref do not store anything, just serialize into LinkGraph directly. If put into block data, it needs to keep some intermediate state so it can create block, then symbols, then go back to edges.

Reply:
Not sure I'm following. Maybe I'll try to write the patch after you land.

Reply:
Fixed.